lucille-core/src/main/java/com/kmwllc/lucille/connector/storageclient/AzureStorageClient.java

```diff
@@ -0,0 +1,186 @@
+package com.kmwllc.lucille.connector.storageclient;
+
+import com.azure.storage.blob.BlobContainerClient;
+import com.azure.storage.blob.BlobContainerClientBuilder;
+import com.azure.storage.blob.models.BlobItem;
+import com.azure.storage.blob.models.BlobItemProperties;
+import com.azure.storage.blob.models.ListBlobsOptions;
+import com.azure.storage.common.StorageSharedKeyCredential;
+import com.kmwllc.lucille.connector.FileConnector;
+import com.kmwllc.lucille.core.ConnectorException;
+import com.kmwllc.lucille.core.Document;
+import com.kmwllc.lucille.core.Publisher;
+import com.typesafe.config.Config;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.commons.io.FilenameUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AzureStorageClient extends BaseStorageClient {
+
+  private BlobContainerClient containerClient;
+  private static final Logger log = LoggerFactory.getLogger(AzureStorageClient.class);
+
+  public AzureStorageClient(URI pathToStorage, String docIdPrefix, List<Pattern> excludes, List<Pattern> includes,
+      Map<String, Object> cloudOptions, Config fileOptions) {
+    super(pathToStorage, docIdPrefix, excludes, includes, cloudOptions, fileOptions);
+  }
+
+  @Override
+  protected String getContainerOrBucketName() {
+    return pathToStorageURI.getPath().split("/")[1];
+  }
+
+  @Override
+  protected String getStartingDirectory() {
+    String path = pathToStorageURI.getPath();
+    String[] subPaths = path.split("/", 3);
```

Comment:
A standard URI for connecting to azure blob client is `https://<storageName>.blob.core.windows.net/<containerName>/folder2/folder3/`. When we want to list the blobs in `folder3` as in this case, we want to feed `folder2/folder3/` as the starting directory to the client.

`pathToStorageURI.getPath()` would return `/<containerName>/folder2/folder3/`. `path.split("/", 3)` would split up the string to return `[, <containerName>, folder2/folder3/]`, splitting up container name so that we are left with `folder2/folder3/` as the starting directory.

Reply:
In the case where URI is `https://<storageName>.blob.core.windows.net/<containerName>/`, then we want to feed "" as the starting directory to the client. `path.split("/", 3)` would split up the string to return `[, <containerName>, ]`, splitting up container name so that we are left with "" as the starting directory. If URI is `https://<storageName>.blob.core.windows.net/<containerName>`, then again we want to return "" as starting directory. `path.split("/", 3)` would return array of length == 2, and so will return ""

Reply:
Ah ok, that makes sense. I was thinking this would only emit the first 3 splits and throw away any text after that, instead of it being part of the 3rd split. Makes sense, thanks!

Reply:
added a code comment for clarification in [2ee666a](https://github.com/kmwtechnology/lucille/pull/217/commits/2ee666a18534e713f42c9a4ab0c4fabd84b5f477)